zaza/openstack/charm_tests/kerberos/setup.py

```diff
@@ -0,0 +1,239 @@
+        domain=domain
+    )
+
+
+def setup_kerberos_configuration_for_test_host():
+    """Retrieve the keytab and krb5.conf to setup the ubuntu test host."""
+    kerberos_server = zaza.model.get_units('kerberos-server')[0]
+    ubuntu_test_host = zaza.model.get_units('ubuntu-test-host')[0]
+
+    dump_file = "krb5.keytab"
+    remote_file = "/etc/krb5.keytab"
+    host_keytab_path = '/home/ubuntu/krb5.keytab'
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        tmp_file = "{}/{}".format(tmpdirname, dump_file)
+        logging.info("Retrieving {} from {}.".format(remote_file,
+                                                     kerberos_server.name))
+        zaza.model.scp_from_unit(
+            kerberos_server.name,
+            remote_file,
+            tmp_file)
+
+        logging.info("SCP {} to {} on {}.".format(tmp_file,
+                                                  host_keytab_path,
+                                                  ubuntu_test_host.name))
+        zaza.model.scp_to_unit(
+            ubuntu_test_host.name,
+            tmp_file,
+            host_keytab_path)
+
+    dump_file = "krb5.conf"
+    remote_file = "/etc/krb5.conf"
+    temp_krb5_path = "/home/ubuntu/krb5.conf"
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        tmp_file = "{}/{}".format(tmpdirname, dump_file)
+        logging.info("Retrieving {} from {}".format(remote_file,
+                                                    kerberos_server.name))
+        zaza.model.scp_from_unit(
+            kerberos_server.name,
+            remote_file,
+            tmp_file)
+
+        logging.info("SCP {} to {} on {}.".format(tmp_file,
+                                                  temp_krb5_path,
+                                                  ubuntu_test_host))
+        zaza.model.scp_to_unit(
+            ubuntu_test_host.name,
+            tmp_file,
+            temp_krb5_path)
+        logging.info('Moving {} to {} on {}.'.format(temp_krb5_path,
+                     remote_file, ubuntu_test_host.name))
+        zaza.model.run_on_unit(ubuntu_test_host.name, ('sudo mv {} {}'.
+                               format(temp_krb5_path, remote_file)))
+
+
+def install_apt_packages_on_ubuntu_test_host():
+    """Install apt packages on a zaza unit."""
+    ubuntu_test_host = zaza.model.get_units('ubuntu-test-host')[0]
+    packages = ['krb5-user', 'python-openstackclient', 'python-pip', 'gcc',
+                'python-dev', 'libkrb5-dev']
+    for package in packages:
+        logging.info('Installing {}'.format(package))
+        result = zaza.model.run_on_unit(ubuntu_test_host.name,
+                                        "apt install {} -y".format(package))
+        assert result['Code'] == '0', result['Stderr']
+
+
+def install_python_packages_on_ubuntu_test_host():
+    """Install python packages on a zaza unit."""
+    ubuntu_test_host = zaza.model.get_units('ubuntu-test-host')[0]
+    package = 'keystoneauth1[kerberos]'
```

Comment:
Is there a reason that we can't use the apt package here too?  If so, we would prefer that here, for the sake of having just one package (dependency) management system in play on the host.

Mixing userspace pip (ie. outside of a venv) with system-level packages is known to invite eventual trouble when dependency requirements clash.

Reply:
I'm not sure actually. The Openstack docs uses pip to install this extra module. https://docs.openstack.org/keystoneauth/latest/api/keystoneauth1.extras.kerberos.html 

I know that python-keystoneauth1 can be installed with apt, but I believe it does not include the kerberos module. I can test and see if it is actually the case.

Reply:
I've tried installing python-keystoneauth1 or python-keystoneauth1[kerberos] instead of pip install and it doesn't work. If there is another way to do this that I'm not away of, please let me know, otherwise I think we need to go forward with the pip install.

Reply:
I think this still needs more work.  AFAICT, python3-keystoneauth1 contains the extras.kerberos bits.  Please evaluate with purely python3 packages via apt.

Reply:
This has been replaced by python3-openstackclient and python3-requests-kerberos.